### PR TITLE
Ese QDoRA instead of just LoRA

### PR DESCRIPTION
**Pull request by strickvl**
Uses QDoRA instead of just LoRA. This has some efficiency benefits, notably faster training without a corresponding performance loss. See [this](https://www.answer.ai/posts/2024-04-26-fsdp-qdora-llama3.html) for more details.

The code on the base branch was also unformatted, but (aside from formatting changes) I've highlighted the specific lines where I've actually changed code.

### CODE REVIEW

File: llm-lora-finetuning/utils/loaders.py [hunk 1]
```diff
@@ -20,7 +20,12 @@
 
 import torch
 from datasets import Dataset
-from peft import LoraConfig, get_peft_model, prepare_model_for_kbit_training
+from peft import (
+    LoraConfig,
+    LoraRuntimeConfig,
+    get_peft_model,
+    prepare_model_for_kbit_training,
+)
```
Comment (strickvl) [inline llm-lora-finetuning/utils/loaders.py:28]: I changed this

File: llm-lora-finetuning/utils/loaders.py [hunk 2]
```diff
@@ -90,6 +98,8 @@ def load_base_model(
             bias="none",
             lora_dropout=0.05,  # Conventional
             task_type="CAUSAL_LM",
+            use_dora=True,
+            runtime_config=LoraRuntimeConfig(ephemeral_gpu_offload=True),
```
Comment (strickvl) [inline llm-lora-finetuning/utils/loaders.py:102]: I added these two lines

**Review by htahir1 — APPROVED** [review]

**Review by avishniakov — APPROVED** [review]
Nice one 👍🏼
